jdbc/src/main/java/com/interface21/jdbc/core/JdbcTemplate.java

```diff
@@ -22,58 +21,64 @@ public JdbcTemplate(final DataSource dataSource) {
+        });
     }
 
-    public <T> T findOne(final String sql, final RowMapper<T> rowMapper, final Object... args) {
-        try (final Connection conn = dataSource.getConnection();
-             final PreparedStatement pstmt = conn.prepareStatement(sql)) {
+    public <T> T queryForObject(final String sql, final RowMapper<T> rowMapper, Object... args) {
+        return queryForObject(sql, createPreparedStatementSetter(args), rowMapper);
+    }
 
-            log.debug("query : {}", sql);
-            setParameters(pstmt, args);
+    public <T> T queryForObject(final String sql, final PreparedStatementSetter pss, final RowMapper<T> rowMapper) {
+        final List<T> results = query(sql, pss, rowMapper);
+        if (results.isEmpty()) {
+            throw new DataAccessException("Incorrect result size: expected 1, actual 0");
+        }
+        return results.getFirst();
+    }
 
-            try (final ResultSet rs = pstmt.executeQuery()) {
-                if (rs.next()) {
-                    return rowMapper.mapRow(rs);
+    public <T> List<T> query(final String sql, final RowMapper<T> rowMapper, Object... args) {
+        return query(sql, createPreparedStatementSetter(args), rowMapper);
+    }
+
+    public <T> List<T> query(final String sql, final PreparedStatementSetter pss, final RowMapper<T> rowMapper) {
+        return execute(sql, pstmt -> {
+            pss.setValues(pstmt);
+            try (ResultSet rs = pstmt.executeQuery()) {
+                final List<T> results = new ArrayList<>();
+                int rowNum = 0;
+                while (rs.next()) {
+                    results.add(rowMapper.mapRow(rs, rowNum++));
                 }
-                return null;
+                return results;
             }
-        } catch (SQLException e) {
-            log.error(e.getMessage(), e);
-            throw new DataAccessException(e);
-        }
+        });
     }
 
-    public <T> List<T> findAll(final String sql, final RowMapper<T> rowMapper, final Object... args) {
+    private <T> T execute(final String sql, final PreparedStatementCallback<T> callback) {
         try (final Connection conn = dataSource.getConnection();
              final PreparedStatement pstmt = conn.prepareStatement(sql)) {
 
             log.debug("query : {}", sql);
-            setParameters(pstmt, args);
 
-            try (final ResultSet rs = pstmt.executeQuery()) {
-                final List<T> results = new ArrayList<>();
-                while (rs.next()) {
-                    results.add(rowMapper.mapRow(rs));
-                }
-                return results;
-            }
+            return callback.doInPreparedStatement(pstmt);
         } catch (SQLException e) {
             log.error(e.getMessage(), e);
             throw new DataAccessException(e);
         }
     }
 
+    private PreparedStatementSetter createPreparedStatementSetter(Object... args) {
+        return pstmt -> setParameters(pstmt, args);
+    }
+
```

Comment:
`setParameters` 메서드가 `PreparedStatementSetter` 한 곳에서만 쓰이고 있네요! 아래처럼 하나의 메서드로 쓸 수 있을 것 같은데 어떻게 생각하시나용?

```suggestion
    private PreparedStatementSetter createPreparedStatementSetter(Object... args) {
        return pstmt -> {
            for (int i = 0; i < args.length; i++) {
                pstmt.setObject(i + 1, args[i]);
            }
        };
    }
```

Reply:
이전 step1에서 사용했던 `setParameters`를 그대로 사용하면서 그런 것 같네요ㅎㅎ 제안해주신 방향으로 수정하겠습니다!